File: gtci/03-sliding-window/python3/4-fruits-into-baskets.py
class Solution:
    def findLength(self, fruits):
        max_length = 0

        window_start = 0
        basket = {}
        for window_end in range(len(fruits)):
            if fruits[window_end] not in basket:
                basket[fruits[window_end]] = 0
            basket[fruits[window_end]] += 1

            if len(basket) <= 2:
                max_length = max(max_length, window_end - window_start + 1)

            while len(basket) > 2:
                basket[fruits[window_start]] -= 1
                if basket[fruits[window_start]] == 0:
                    del basket[fruits[window_start]]

                window_start += 1

        return max_length


if __name__ == "__main__":
    sol = Solution()
    assert sol.findLength(["A", "B", "C", "A", "C"]) == 3
    assert sol.findLength(["A", "B", "C", "B", "B", "C"]) == 5
    assert sol.findLength(["A", "A", "A", "A"]) == 4

    print("All test cases passed.")

File: gtci/03-sliding-window/cpp/4-fruits-into-baskets.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
   public:
    int findLength(const vector<char> &fruits) {
        int maxLength = 0;

        int windowStart = 0;
        std::unordered_map<char, int> basket;
        for (int windowEnd = 0; windowEnd < fruits.size(); windowEnd++) {
            basket[fruits[windowEnd]] += 1;

            if (basket.size() <= 2)
                maxLength = std::max(maxLength, windowEnd - windowStart + 1);

            while (basket.size() > 2) {
                basket[fruits[windowStart]] -= 1;
                if (basket[fruits[windowStart]] == 0)
                    basket.erase(fruits[windowStart]);

                windowStart++;
            }
        }

        return maxLength;
    }
};

int main() {
    Solution sol;

    assert(sol.findLength(vector<char>{'A', 'B', 'C', 'A', 'C'}) == 3);
    assert(sol.findLength(vector<char>{'A', 'B', 'C', 'B', 'B', 'C'}) == 5);
    assert(sol.findLength(vector<char>{'A', 'A', 'A', 'A'}) == 4);

    std::cout << "All test cases passed." << std::endl;
    return 0;
}
